Dockerfile

```diff
@@ -2,37 +2,113 @@
+# hadolint ignore=DL3006
+FROM ${BASE_IMAGE}@${BASE_DIGEST_ARM64} as base-arm64
+ARG BASE_DIGEST_ARM64
+ARG BASE_DIGEST="${BASE_DIGEST_ARM64}"
+
+### Base image ##
+# All package installation, updates, etc., anything with APT should be done here in a single step
+# hadolint ignore=DL3006
+FROM base-${TARGETARCH} as base
+WORKDIR /
+
+# Upgrade all outdated packages and install missing ones (e.g. locales, tini)
+# This breaks reproducibility of builds, but is acceptable to gain access to security patches faster
+# than the base image releases
+# FYI, installing packages via APT also updates the dpkg files, which are few MBs, but removing or
+# caching them could break stuff (like not knowing the package is present) or container scanners
+# hadolint ignore=DL3008
 RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
     --mount=type=cache,target=/var/lib/apt,sharing=locked \
+    --mount=type=cache,target=/var/log/apt,sharing=locked \
     apt-get -qq update && \
-    apt-get install -y --no-install-recommends tini=0.19.0-1 && \
-    cp /usr/bin/tini .
-COPY --link --chown=1000:0 docker/utils/startup.sh .
+    apt-get install -yqq --no-install-recommends tini ca-certificates && \
+    apt-get upgrade -yqq --no-install-recommends
+
+
+### AMD64 JDK image ###
+# JDK_DIGEST_AMD64 is defined at the top of the Dockerfile
+# hadolint ignore=DL3006
+FROM ${JDK_IMAGE}@${JDK_DIGEST_AMD64} as jdk-amd64
+
+### ARM64 JDK image ##
+# JDK_DIGEST_ARM64 is defined at the top of the Dockerfile
+# hadolint ignore=DL3006
+FROM ${JDK_IMAGE}@${JDK_DIGEST_ARM64} as jdk-arm64
+
+### Build custom JRE using the base JDK image
+# hadolint ignore=DL3006
+FROM jdk-${TARGETARCH} as jre-build
+
+# Build a custom JRE which will strip down and compress modules to end up with a smaller Java \
+# distribution than the official JRE. This will also include useful debugging tools like
+# jcmd, jmod, jps, etc., which take little to no space. Anecdotally, compressing modules add around
+# 10ms to the start up time, which is negligible considering our application takes ~10s to start up.
+# See https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/
+# hadolint ignore=DL3018
+RUN jlink \
+     --add-modules ALL-MODULE-PATH \
+     --strip-debug \
+     --no-man-pages \
+     --no-header-files \
+     --compress=2 \
+     --output /jre && \
+   rm -rf /jre/lib/src.zip
+
+### Java base image
+FROM base AS java
+WORKDIR /
+
+# Inherit from previous build stage
+ARG JAVA_HOME=/opt/java/openjdk
+
+# Default to UTF-8 file encoding
+ENV LANG='C.UTF-8' LC_ALL='C.UTF-8'
+
+# Setup JAVA_HOME and binaries in the path
+ENV JAVA_HOME ${JAVA_HOME}
+ENV PATH $JAVA_HOME/bin:$PATH
+
+# Copy JRE from previous build stage
+COPY --from=jre-build /jre ${JAVA_HOME}
+
+# https://github.com/docker-library/openjdk/issues/212#issuecomment-420979840
+# https://openjdk.java.net/jeps/341
+# TL;DR generate some class data sharing for faster load time
+RUN java -Xshare:dump;
```

Comment:
This is cool! We could probably try and run the broker/gateway here and benefit even for from CDS?

Reply:
I tried that but didn't see much difference locally :(

[See this comment](https://github.com/camunda/zeebe/issues/12992#issuecomment-1586740682) hidden in a random issue :smile: 

But I have faith you could make it work ;)

Reply:
Tried it again today, and I could get the Spring startup down from 3-4 seconds to < 2 seconds, so I guess it does sort of work :sweat_smile: 

You can try it yourself, see if you can replicate the same results. First, you need to remove the `config` directory from the class path. That will trip up creating the CDS archive unfortunately, since it doesn't contain Java code. Maybe there's a workaround, but I'm not aware of it. 

Then repackage the `dist` module, then just go into `dist/target/camunda-zeebe`. From there:

```shell
JAVA_OPTS="-Xshare:off -XX:DumpLoadedClassList=broker.classlist" bin/broker
JAVA_OPTS="-Xshare:dump -XX:SharedArchiveFile=broker.jsa -XX:SharedClassListFile=broker.classlist" bin/broker
JAVA_OPTS="-XX:SharedArchiveFile=broker.jsa" bin/broker
```

With a static archive like this though, there's a couple thing:

1. The archive is sort of big (92MB for me), but I guess if it cuts down startup time by half, that's worth it
2. It seems _very_ sensitive to the class path, so I'd have to check if changing the class path later might not break stuff

That said, starting with Java 19 (and Java 21 is coming in September so...) you can just do:

```shell
JAVA_OPTS="-XX:+AutoCreateSharedArchive -XX:SharedArchiveFile=/tmp/broker.jsa" bin/broker
```

And it will auto magically create everything for you. It'll also create a slightly smaller archive as it won't include core classes.

So I would recommended we wait for the Java 21 upgrade (which we should 100% schedule soon :smile:), then apply this. But yeah, it cuts down the startup from 3-5s to <2s locally for me, so I think it's definitely worth doing. Not sure exactly what I did in my last test that it wasn't working :smile:

Reply:
The startup times I was referring to are the Spring Boot ones. Zeebe itself will add some variable time obviously.

Without AppCDS:

![image](https://github.com/camunda/zeebe/assets/43373/9d60a3d4-00fe-4807-85d0-8be641e3eaec)


With AppCDS:

![image](https://github.com/camunda/zeebe/assets/43373/7458fd90-bf47-4389-b86f-1b35cfe61d50)

Reply:
I've got [a branch based on this one which also has an AppCDS generating stage](https://github.com/camunda/zeebe/blob/np-docker-appcds/Dockerfile), and the corresponding image, [gcr.io/zeebe-io/zeebe:static-appCds](gcr.io/zeebe-io/zeebe:static-appCds)